src/Runner.ts

```diff
@@ -153,145 +224,96 @@ function parseHtml(content: string): { css: string, body: string, scripts: strin
-			iframe.contentWindow.location.reload();
-		});
-	}
-
-	/**
-	 * Generate the document
-	 * @param param0 The options to use
-	 */
-	getDoc({ css = [], bodyAttributes = {}, dependencies, loaderSrc, html = '', modules, scripts = [] }: GetDocOptions): string {
-		return docSrc`<!DOCTYPE html>
-			<html>
-			<head>
-				${scripts}
-				${css}
-			</head>
-			<body${bodyAttributes}>
-				${html}
-				<script src="${loaderSrc}"></script>
-				<script>require.config({
-	paths: ${dependencies},
-	packages: ${getPackages(dependencies)}
-});
-
-var cache = {};
-//# sourceURL=web-editor/config.js
-				</script>
-				${modules}
-				<script>require.cache(cache);
-/* workaround for dojo/loader#124 */
-require.cache({});
-
-require([ 'tslib', '@dojo/core/request', '../support/providers/amdRequire' ], function () {
-	var request = require('@dojo/core/request').default;
-	var getProvider = require('../support/providers/amdRequire').default;
-	request.setDefaultProvider(getProvider(require));
-	require([ 'src/main' ], function () { });
-});
-//# sourceURL=web-editor/bootstrap.js
-				</script>
-			</body>
-			</html>`;
-	}
-
-	/**
-	 * Get the emit from the current project and run it in the runner's `iframe`
-	 */
-	async run(): Promise<void> {
-		if (!project.isLoaded()) {
-			throw new Error('Project not loaded.');
+	@afterRender()
+	public updateSource(node?: DNode): DNode | undefined {
+		if (!this._iframe) {
+			return node;
 		}
-
-		const program = await project.emit();
-
-		const modules = program
-			.filter(({ type }) => type === ProjectFileType.JavaScript || type === ProjectFileType.SourceMap)
-			.reduce((map, { name, text, type }) => {
-				const mid = name.replace(/\.js(?:\.map)?$/, '');
-				if (!(mid in map)) {
-					map[mid] = { code: '', map: '' };
-				}
-				map[mid][type === ProjectFileType.JavaScript ? 'code' : 'map'] = text;
-				return map;
-			}, {} as { [mid: string]: { code: string; map: string; } });
-
-		const css = program
-			.filter(({ type }) => type === ProjectFileType.CSS)
-			.map(({ name, text }) => { return { name, text }; });
-
-		const dependencies = project.getDependencies();
-
-		const { css: text, body: html, scripts } = parseHtml(project.getIndexHtml());
-		if (text) {
-			css.unshift({ name: 'project index', text });
+		if (this._updating) {
+			return node;
 		}
+		if (this.properties.modules) {
+			this._updating = true;
+			const source = getSource(this.properties);
+			writeIframeDoc(this._iframe, source, this._onIframeError)
+				.then(() => {
+					this._updating = false;
+					const { onRun } = this.properties;
+					onRun && onRun();
+				});
+		}
+		return node;
+	}
 
-		const source = this.getDoc({
-			css,
-			html,
-			dependencies,
-			loaderSrc: 'https://unpkg.com/@dojo/loader/loader.min.js',
-			modules,
-			scripts
-		});
-		await this._writeIframeDoc(source);
+	public render() {
+		return v('div', {
+			classes: this.classes(css.base)
+		}, [ v('iframe', {
```

Comment:
It means you don't need to "get" the element from widget system - which could be advantageous once those hooks have been removed.

Perhaps could then restructure the way the "afterCreate", "afterUpdate" functionality works also.